### ISSUE TITLE
HashWithIndifferentAccess#compact returns Hash in Ruby 2.4

### ISSUE THREAD

**Issue by rwz** [rails/rails#28116]
Tested with ActiveSupport 4.2.8.

In Ruby 2.3.3

```ruby
ActiveSupport::HashWithIndifferentAccess.new.compact.class # => ActiveSupport::HashWithIndifferentAccess
```

In Ruby 2.4.0

```ruby
ActiveSupport::HashWithIndifferentAccess.new.compact.class # => Hash
```

Edit: It seems to work correctly in ActiveSupport 5.0.1 though.

**Comment by rafaelfranca** [comment]
Fixed by 828e0199b79a23e2b098933dfc117b2ac21861a1

**Comment by rwz** [comment]
Thanks
